### DISCUSSION TITLE
Hammer Db on Multiple Pluggable Databases

### DISCUSSION THREAD

**Discussion by sachinpadha** [TPC-Council/HammerDB#239] [General]
Hi Steve,

We have one Database server and that contains multiple PDBs. So it means it has different TNS entries. In our case we have 5 PDBs with 2 nodes each.

I executed the Work load one PDB and it worked and now I want to run a workload on 5 PDBs, how can I achieve that? Because config/oracle.xml contains only one TNS entry. Please help in this regard. Also, in config/connectpool/oracpool.xml I have given the multiple connection entries but while using buildschema it is build only in one Database and not on another database. I want to do this operation at same movement (parallely)

**Comment by sm-shaw** [comment]
Hi,
For the schema build by default it will support building 1 schema at a time.  It is creating and inserting the data at the same time, so it makes sense to focus the build resources on 1 database only. However, if you want to support multiple database builds then you can consider using the Datagen feature to create the data for loading databases manually in bulk, scripting with the CLI or using the primary/replica feature to drive multiple GUI or CLI clients to run the same workload including builds. 
For the test you have 2 options, One solution is to use the connect pool option for one instance of HammerDB to connect to multiple Oracle instances although by default it captures performance data from only the main instance, this would be typical for Oracle RAC where you're testing multiple instances of the same clustered database and this is what that feature was designed for.  However, because some databases will not support cluster wide transaction statistics this was already thought of and when using Connect Pooling HammerDB will report client side transaction data as well. Therefore, you can use this functionality for testing multiple unrelated instances, discard the main NOPM/TPM report and view the client side stats.  
Alternatively  you can use the primary/replica feature to create multiple instances of HammerDB and drive them with the same load, you can do this like the build in both the GUI and CLI. This is typical and was designed for testing multiple virtualized instances on the same host at the same time originally in the GUI.

**Reply by sachinpadha** [reply]
Thank you very much for your detailed reply. I will do below steps:

1) I will buildschema on one Database and will do clone to other database using pluggable feature.
2) While executing the workload through CLI, I will use "primary" SWITCHMODE but in CLI there is some issue:

switchmode Primary --- This I will write in my Primary DB of CLI custom script.
switchmode Replica 12345 AnotherInstance --- This command I will write on another Database of CLI custom script.

But question is .. In CLI switchmode of Primary DB, how can I capture the Primary id because it is dynamically generated. How can I achieve this through CLI ? If you

Also I want to run the entire workload may be for 48hrs. Currently I am using 100 Virtual users with 500 Warehouses and it ran for 4 hrs.

If you any example to achieve switchmode via CLI and run the workload for 48hrs please let me know

**Comment by sm-shaw** [comment]
Yes, this is built in with the Variable Workloads https://www.hammerdb.com/docs/ch04s06.html#d0e1901
With this you define multiple steps in XML and then the replicas connect automatically to the primary - exactly what you want. Therefore, you can define a special case of step workload where they start 1 minute after each other and all finish at the same time. 
If you want to run for a long period of time make sure you increase the number of transactions per virtual user - the default is 1000000, so the Virtual User will end when it reaches this number of transactions. You can increase this to any value you like.

**Reply by sachinpadha** [reply]
Below is my config/oracle.xml:

<?xml version="1.0" encoding="utf-8"?>
<oracle>
    <connection>
        <system_user>tpcc</system_user>
        <system_password>tpcc</system_password>
        <instance>scanIP:1521/primarydb</instance>
        <rac>0</rac>
    </connection>
    <tpcc>
        <schema>
            <count_ware>500</count_ware>
	    <num_vu>100</num_vu>
            <tpcc_user>tpcc</tpcc_user>
            <tpcc_pass>tpcc</tpcc_pass>
            <tpcc_def_tab>tpcc</tpcc_def_tab>
            <tpcc_ol_tab>tpcc</tpcc_ol_tab>
            <tpcc_def_temp>temp</tpcc_def_temp>
	    <partition>false</partition>
	    <hash_clusters>false</hash_clusters>
	    <tpcc_tt_compat>false</tpcc_tt_compat>
        </schema>
        <driver>
            <total_iterations>1000000</total_iterations>
            <raiseerror>false</raiseerror>
            <keyandthink>false</keyandthink>
	    <checkpoint>false</checkpoint>
            <ora_driver>timed</ora_driver>
            <rampup>2</rampup>
            <duration>5</duration>
            <allwarehouse>true</allwarehouse>
            <ora_timeprofile>false</ora_timeprofile>
	    <async_scale>false</async_scale>
	    <async_client>10</async_client>
	    <async_verbose>false</async_verbose>
	    <async_delay>1000</async_delay>
	    <connect_pool>xml</connect_pool>
        </driver>
    </tpcc>
    <tpch>
        <schema>
            <scale_fact>1</scale_fact>
            <tpch_user>tpch</tpch_user>
            <tpch_pass>tpch</tpch_pass>
            <tpch_def_tab>tpchtab</tpch_def_tab>
            <tpch_def_temp>temp</tpch_def_temp>
            <num_tpch_threads>1</num_tpch_threads>
	    <tpch_tt_compat>false</tpch_tt_compat>
        </schema>
        <driver>
            <total_querysets>1</total_querysets>
            <raise_query_error>false</raise_query_error>
            <verbose>false</verbose>
            <degree_of_parallel>2</degree_of_parallel>
	    <refresh_on>false</refresh_on>
	    <update_sets>1</update_sets> 
	    <trickle_refresh>1000</trickle_refresh> 
	    <refresh_verbose>false</refresh_verbose> 
	    <cloud_query>false</cloud_query>
        </driver>
    </tpch>
</oracle>


In the above XML, I have changed the connect_pool to xml. Is this fine?
And in config/connectpool/ , the below tags are incorrect: Can you please provide c1 connection for Oracle?
  <c1>
  	<tpcc_user>tpcc</tpcc_user>
        <tpcc_pass>tpcc</tpcc_pass>
        <instance>oracle1</instance>
    </c1>

Once all these settings are done I will execute below custom script from Primary DB only and no need to execute it from other DB or any other nodes? Is my understanding correct? or Do I need to execute the custom script on all Nodes and all Instances.? Can you verify if Oracle.xml, oracpool.xml and below custom scripts are fine? All these are working fine on single node.


#!/usr/bin/tclsh
proc runtimer { seconds } {
set x 0
set timerstop 0
while {!$timerstop} {
incr x
after 1000
  if { ![ expr {$x % 60} ] } {
          set y [ expr $x / 60 ]
          puts "Timer: $y minutes elapsed"
  }
update
if {  [ vucomplete ] || $x eq $seconds } { set timerstop 1 }
    }
return
}
puts "SETTING CONFIGURATION"
dbset db ora
diset tpcc ora_driver timed
diset tpcc ora_rampup 2
diset tpcc ora_duration 5
vuset logtotemp 1
switchmode Primary
loadscript
puts "SEQUENCE STARTED"
foreach z { 96 100 104 } {
puts "$z VU TEST"
vuset vu $z
vucreate
vurun
runtimer 600
vudestroy
after 5000
}
puts "TEST SEQUENCE COMPLETE"

**Comment by sm-shaw** [comment]
There is a blog post here that can help you using the CLI https://www.hammerdb.com/blog/uncategorized/hammerdb-cli-101/

**Reply by sachinpadha** [reply]
I have already checked the blog but there is no mentioned of xml pool via CLI script and also I need one correct sample for config/connectpool/oracpool.xml file. Because the default sample is not serving the purpose. So please help us in this regard.
Also question remains the same: I need to execute custom script on all Instances or only in Primary instance?

**Reply by sm-shaw** [reply]
As it is not a cost based service unfortunately the HammerDB developers are not able to give extended support in each and every case such as checking user scripts, configs etc.
Please note that these are advanced features and where possible documentation and blogs are provided to give guides on how to use them.  This is the guide to Connect Pooling https://www.hammerdb.com/blog/uncategorized/hammerdb-v4-0-new-features-pt4-connect-pooling-for-clusters/. A guide to using the step workloads is in progress.
